### PR TITLE
Migrate to `pydantic>=2.5`, `torch==2.1.1`, `ray[default]>=2.9`

### PR DESCRIPTION
**Pull request by spillai**
## Summary
 - upgrades to pydantic, torch, ray and agi-pack to support `pydantic>2.5` upgrade
 - Ray requires an upgrade to `2.9` for pydantic>2 update: https://github.com/ray-project/ray/issues/39722
 - Updated `agi-pack` with `pydantic>2` support: https://github.com/spillai/agi-pack/pull/16
 - various migration upgrades to ModelSpec and other dataclasses with new pydantic `field_validator` syntax

## Related issues

#220 

## Checks

- [x] `make lint`: I've run `make lint` to lint the changes in this PR.
- [x] `make test`: I've made sure the tests (`make test-cpu` or `make test`) are passing.
- Additional tests:
   - [x] Benchmark tests (when contributing new models)
   - [x] GPU/HW tests

### CODE REVIEW

File: nos/common/spec.py [hunk 1]
```diff
@@ -151,22 +149,42 @@ class FunctionSignature:
     method: str
     """Class method name. (e.g. forward, __call__ etc)"""
 
-    init_args: Tuple[Any, ...] = field(default_factory=tuple)
```
Comment (outtanames) [inline nos/common/spec.py:154]: [nit] might be good to validate `method != None`? same for `func_or_cls`

File: nos/managers/model.py [hunk 2]
```diff
@@ -237,7 +237,7 @@ def _get_actor_options(cls, spec: ModelSpec, deployment: ModelDeploymentSpec) ->
 
         if spec.runtime_env is not None:
             logger.debug("Using custom runtime environment, this may take a while to build.")
-            actor_opts["runtime_env"] = RuntimeEnv(**asdict(spec.runtime_env))
+            actor_opts["runtime_env"] = RuntimeEnv(**spec.runtime_env.model_dump())
```
Comment (outtanames) [inline nos/managers/model.py:240]: ?

File: tests/common/test_common_spec.py [hunk 3]
```diff
@@ -238,7 +238,7 @@ def txt2img(self, texts: str) -> Image.Image:
 
     # Image generation (txt2img)
     txt2img_signature = FunctionSignature(
-        Custom,
+        func_or_cls=Custom,
```
Comment (outtanames) [inline tests/common/test_common_spec.py:241]: 'func_or_class' is kind of a confusing name given that we also have method.

**Review by outtanames — APPROVED** [review]

File: nos/common/spec.py (hunk 1 above)
Comment (spillai) [reply nos/common/spec.py:154]: That is already handled in the new pydantic validation

**Pull request merged**